lib/image_optim/cache.rb

```diff
@@ -0,0 +1,64 @@
+require 'digest/sha1'
+require 'fspath'
+
+class ImageOptim
+  # Handles image cache
+  class Cache
+    def initialize(image_optim, workers_by_format)
+      return unless image_optim.cache_dir
+      @cache_dir = FSPath.new(image_optim.cache_dir)
+      @cache_worker_digests = image_optim.cache_worker_digests
+      @options_by_format = Hash[workers_by_format.map do |format, workers|
+        [format, workers.map(&:inspect).sort.join(', ')]
+      end]
+      @bins_by_format = Hash[workers_by_format.map do |format, workers|
+        [format, workers.map(&:used_bins).flatten!.map! do |sym|
+          bin = image_optim.resolve_bin!(sym)
+          "#{bin.name}[#{bin.digest}]"
+        end.sort!.uniq.join(', ')]
+      end]
+    end
+
+    def fetch(original)
+      return yield unless @cache_dir
+
+      cached = @cache_dir / digest(original)
+      return cached if File.file?(cached)
+      return if File.symlink?(cached)
+
+      optimized = yield
+
+      digest = digest(original)
+      cached = @cache_dir / digest
```

Comment:
`cached` is already assigned previously, also `digest` on previous line is also not needed